### PR TITLE
Make parent connection automatic (closes #294)

### LINKED ISSUE #294: Make parent-child connection always coherent
We may want to ensure the PSyIR parent-child connection is always coherent. Simply adding to Node `__init__` or `addChild` methods something like:
```
self._children.append(child)
if child.parent is not self:
    child.parent = self
```
breaks the unittest

### PR DESCRIPTION
**Pull request by sergisiso**
This PR makes the parent setter unnecessary as the connection will already be set up when the node is added as a child (from addchild or any direct list operation).

It solves the remaining issue when checking that children additions shouldn't be duplicated even if the parent is the same. e.g. `node.children = [node2, node2]` will now fail.

There is still one way to set a parent link without the complementing child connection which is with the constructor:
```
node1 = Node(parent=node2)
node2.addchild(node1)
```
These are allowed, but if the child connection doesn't match what was stated in the first statement, it will fail.

The parent argument can be omited most of the time, it is only necessary if the constructor or incomplete node needs to check the scoping symbol table where it belongs. This is necessary mainly for the top-down approach of the fparser2 frontend, and changing it would require a big refactor of the front-end (which I don't think is worth doing).

### CODE REVIEW

**Comment by sergisiso** [comment]
@rupertford @arporter This is ready for the first review.

File: src/psyclone/transformations.py [hunk 1]
```diff
@@ -1960,29 +1928,21 @@ def apply(self, outer, options=None):
         schedule = outer.root
         inner = outer.loop_body[0]
         parent = outer.parent
+        index = parent.children.index(outer)
```
Comment (arporter) [inline src/psyclone/transformations.py:1931]: pylint says `index` is unused.

File: src/psyclone/transformations.py [hunk 2]
```diff
@@ -556,21 +544,18 @@ def __init__(self):
...
         transformation.
 
-        :param parent: the parent Node of the new directive Node.
-        :type parent: :py:class:`psyclone.psyir.nodes.Node`
         :param children: list of child nodes of the new directive Node.
         :type children: list of :py:class:`psyclone.psyir.nodes.Node`
         :param int collapse: number of nested loops to collapse or None if \
                              no collapse attribute is required.
         '''
         from psyclone.psyGen import ACCLoopDirective
```
Comment (arporter) [inline src/psyclone/transformations.py:557]: Please could you move this import to the top.

File: src/psyclone/transformations.py [hunk 3]
```diff
@@ -435,8 +424,7 @@ def _directive(self, parent, children, collapse=None):
                 "directives.")
 
         from psyclone.psyGen import OMPDoDirective
```
Comment (arporter) [inline src/psyclone/transformations.py:426]: Please could you move this import to the top (if possible)

**Review by arporter — CHANGES_REQUESTED** [review]
Looks really good Sergi.
I have a question about when we allow the parent of a node to be over-written but apart from that it's just minor things.
I've checked all files for pylint/pycodestyle.
Please could you update the appropriate documentation with a few words on how the consistency of the parent-child relationship is maintained.

File: src/psyclone/psyGen.py [hunk 4]
```diff
@@ -2460,19 +2462,19 @@ class Kern(Statement):
                  the Algorithm layer code.
     :type call: :py:class:`psyclone.parse.algorithm.KernelCall`
     :param str name: the name of the routine being called.
-    :param arguments: object holding information on the kernel arguments, \
+    :param ArgumentsClass: class to create the kernel arguments, \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:2465]: Done

File: src/psyclone/psyGen.py [hunk 5]
```diff
@@ -2914,13 +2915,12 @@ def lower_to_language_level(self):
...
-        self.parent.children[self.position] = call_node
-        call_node.parent = self.parent
-
+        # Create argument expressions as children of the new node
         for argument in self.arguments.psyir_expressions():
             call_node.addchild(argument)
-            argument.parent = call_node
+
+        # Swap itself with the appropriate Call node
+        self.parent.children[self.position] = call_node
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:2923]: Yes

File: src/psyclone/psyir/nodes/array_member.py [hunk 6]
```diff
@@ -54,7 +54,7 @@ class ArrayMember(ArrayMixin, Member):
     _text_name = "ArrayMember"
 
     @staticmethod
-    def create(member_name, indices, parent=None):
+    def create(member_name, indices):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/array_member.py:57]: Oops, removed.

File: src/psyclone/psyir/nodes/loop.py [hunk 7]
```diff
@@ -531,7 +526,8 @@ def gen_code(self, parent):
         :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
 
         '''
-        from psyclone.psyGen import zero_reduction_variables
+        # Avoid circular dependency
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/loop.py:529]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 8]
```diff
@@ -147,24 +147,48 @@ def _check_is_orphan(self, item):
...
                 " an orphan. It already has a '{2}' as a parent.".format(
                     item.coloured_name(False),
                     self._node_reference.coloured_name(False),
                     item.parent.coloured_name(False)))
 
+    def _set_parent_link(self, item):
+        '''
+        Set parent connection of the given item to this ChildrenList's node.
+
+        :param item: item which the parent connection need to be updated.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:162]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 9]
```diff
@@ -147,24 +147,48 @@ def _check_is_orphan(self, item):
...
+        :type item: :py:class:`psyclone.psyir.nodes.Node`
+        '''
+        if item.parent and item.has_constructor_parent:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:165]: This was right, we allow it to set a parent unless the constructor said it had to be another parent. Once the constructor 
specified parent is finally connected it behaves as a normal node. 

But I decided to move the check into the is_orphan method and add the raises docstring that it was missing. Hopefully this makes things clearer.

File: src/psyclone/psyir/nodes/node.py [hunk 10]
```diff
@@ -147,24 +147,48 @@ def _check_is_orphan(self, item):
...
+        # pylint: disable=protected-access
+        item._parent = self._node_reference
+        item._has_constructor_parent = False
+
+    @staticmethod
+    def _del_parent_link(item):
+        '''
+        Delete parent connection of the given item.
+
+        :param item: item which the parent connection need to be updated.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:183]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 11]
```diff
@@ -147,24 +147,48 @@ def _check_is_orphan(self, item):
...
+        :type item: :py:class:`psyclone.psyir.nodes.Node`
+        '''
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:185]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 12]
```diff
@@ -147,24 +147,48 @@ def _check_is_orphan(self, item):
...
+        # This is done from here with protected access because its the parent
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:186]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 13]
```diff
@@ -147,24 +147,48 @@ def _check_is_orphan(self, item):
...
+        # who it is in charge of maintaining its children connections.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:187]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 14]
```diff
@@ -304,6 +337,9 @@ def __init__(self, ast=None, children=None, parent=None, annotations=None):
                                       self._children_valid_format)
         if children:
             self._children.extend(children)
+        self._has_constructor_parent = False
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:340]: Good idea, and done

File: src/psyclone/psyir/nodes/node.py [hunk 15]
```diff
@@ -804,14 +840,12 @@ def children(self, my_children):
...
+        :raises TypeError: if the given children parameter is not a list.
         '''
-        if my_children is None:
-            self._children = None
-        elif isinstance(my_children, list):
+        if isinstance(my_children, list):
+            self.pop_all_children()  # First remove existing children if any
             self._children = ChildrenList(self, self._validate_child,
                                           self._children_valid_format)
             self._children.extend(my_children)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:851]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 16]
```diff
@@ -821,11 +855,20 @@ def children(self, my_children):
...
         return self._parent
 
-    @parent.setter
-    def parent(self, my_parent):
-        self._parent = my_parent
+    @property
+    def has_constructor_parent(self):
+        '''
+        :returns: whether the constructor has predefined a parent connection
+            which is not yet realised.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:868]: Extended the comment, hopefully it is clearer now.

File: src/psyclone/psyir/nodes/psy_data_node.py [hunk 17]
```diff
@@ -299,21 +299,22 @@ def gen_code(self, parent, options=None):
             be added to each variable name in the post_var_list.
 
         '''
+        from psyclone.psyGen import Kern, InvokeSchedule
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/psy_data_node.py:302]: Done

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 18]
```diff
@@ -342,7 +342,8 @@ def apply(self, node, options=None):
         iloop = Loop.create(i_loop_symbol, lower_bound, upper_bound, step,
                             [assign, jloop])
         # Add the new code to the PSyIR
-        iloop.parent = assignment.parent
         assignment.parent.children.insert(assignment.position, iloop)
         # remove the original matmul
         assignment.parent.children.remove(assignment)
+
+        return iloop.root
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:349]: It was not needed, I reverted this change.

File: src/psyclone/tests/domain/gocean/goloop_test.py [hunk 19]
```diff
@@ -59,8 +59,8 @@ def test_goloop_no_parent():
     goloop = GOLoop(loop_type="inner", parent=schedule)
     schedule.children = [goloop]
     # Now remove parent and children
-    goloop.parent = None
-    goloop.children = None
+    goloop.detach()
+    goloop.children = []
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/gocean/goloop_test.py:63]: Not necessarily, assigning an empty list already takes care of properly detaching all children (by the internal pop_all_children() on the node children setter).

I added a new check in the children setter test to guarantee this behavior.

File: src/psyclone/tests/domain/gocean/transformations/gocean1p0_transformations_test.py [hunk 20]
```diff
@@ -221,10 +221,9 @@ def test_omp_parallel_loop(tmpdir):
 
     omp = GOceanOMPParallelLoopTrans()
     cbtrans = GOConstLoopBoundsTrans()
-    omp_sched, _ = omp.apply(schedule.children[0])
-    newsched, _ = cbtrans.apply(omp_sched, {"const_bounds": True})
+    omp.apply(schedule.children[0])
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/gocean/transformations/gocean1p0_transformations_test.py:224]: Done

File: src/psyclone/tests/domain/gocean/transformations/gocean1p0_transformations_test.py [hunk 21]
```diff
@@ -556,13 +554,11 @@ def test_omp_region_retains_kernel_order3(tmpdir):
     ompl = GOceanOMPLoopTrans()
 
     # Put an OMP Do around the 2nd loop of the schedule
-    omp_schedule, _ = ompl.apply(schedule.children[1])
+    ompl.apply(schedule.children[1])
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/gocean/transformations/gocean1p0_transformations_test.py:557]: Done

File: src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py [hunk 22]
```diff
@@ -170,11 +168,13 @@ def test_kernel():
     assert not nemo_loop.kernel
 
     schedule = nemo_loop.loop_body
+    children = [Assignment.create(Reference(x_var), Literal("3.0", REAL_TYPE))]
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:171]: It wasn't, I removed it.

File: src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py [hunk 23]
```diff
@@ -70,9 +70,7 @@ def test_instance_args():
 
     '''
     variable = DataSymbol("nemo_symbol", INTEGER_TYPE)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:72]: Removed

File: src/psyclone/tests/psyir/nodes/array_reference_test.py [hunk 24]
```diff
@@ -56,8 +56,8 @@ def test_array_node_str():
     array_type = ArrayType(INTEGER_SINGLE_TYPE, [ArrayType.Extent.ATTRIBUTE])
     symbol = DataSymbol("aname", array_type)
     kschedule.symbol_table.add(symbol)
-    assignment = Assignment(parent=kschedule)
-    array = ArrayReference(symbol, parent=assignment)
+    assignment = Assignment()
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/nodes/array_reference_test.py:59]: Removed

File: src/psyclone/tests/psyir/nodes/node_test.py [hunk 25]
```diff
@@ -995,3 +982,80 @@ def test_detach():
 
     # Executing it again still succeeds
     assert node1.detach() is node1
+
+
+def test_parent_references_coherency():
+    ''' Check that the parent references keep updated with the children
+    node operations. '''
+    parent = Schedule()
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/nodes/node_test.py:990]: I think this is already provided by the test_children_is_orphan_validation for all operations?

File: src/psyclone/tests/psyir/nodes/range_test.py [hunk 26]
```diff
@@ -83,23 +81,16 @@ def test_range_init(parser):
 
 def test_range_create():
     ''' Check that the Range.create() method behaves as intended. '''
-    parent = Node()
     start = Literal("1", INTEGER_SINGLE_TYPE)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/nodes/range_test.py:84]: Removed

File: src/psyclone/transformations.py [hunk 27]
```diff
@@ -1960,29 +1928,21 @@ def apply(self, outer, options=None):
...
-        inner.parent = parent
-        parent.children.insert(index, inner)
-        outer.loop_body.children.remove(inner)
+        # Detach the inner code
+        inner_loop_body = inner.loop_body.detach()
 
-        # Move inner's schedule to outer
-        outer.children[3] = inner.loop_body.detach()
-        outer.loop_body.parent = outer
+        # Swap the loops
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/transformations.py:1939]: :)

File: src/psyclone/transformations.py (hunk 1 above)
Comment (sergisiso) [reply src/psyclone/transformations.py:1931]: Removed

File: src/psyclone/transformations.py (hunk 2 above)
Comment (sergisiso) [reply src/psyclone/transformations.py:557]: Done

File: src/psyclone/transformations.py (hunk 3 above)
Comment (sergisiso) [reply src/psyclone/transformations.py:426]: Done

**Comment by sergisiso** [comment]
@aporter I addressed the comments and brought it to master. This is ready for the next review.

File: src/psyclone/psyir/nodes/node.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:163]: Ah I see. That makes sense and I think your updates to the docstring make it clearer.

File: src/psyclone/psyir/nodes/node.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:190]: I think you've missed this one (in `_del_parent_link`)?

File: src/psyclone/psyir/nodes/node.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py]: Ah yes, I get it now :-)

File: src/psyclone/tests/domain/gocean/goloop_test.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/gocean/goloop_test.py:63]: Great.

File: src/psyclone/tests/psyir/nodes/node_test.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/node_test.py:1000]: Oh yes, so it is.

**Review by arporter — CHANGES_REQUESTED** [review]
Almost there now on the code front - just a couple of minor things.

I think we need a new (probably short) section in the PSyIR part of the Developer Guide describing the rules for parent/child relationships and how consistency is maintained. e.g. that the parent property can either be set via the constructor or is set-up by the various child-manipulation methods, the fact that only orphans can be added as children of a node etc. etc.

File: src/psyclone/psyir/nodes/node.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:190]: Done now, and renamed item to node.

File: src/psyclone/tests/psyir/transformations/loop_trans_test.py [hunk 28]
```diff
@@ -71,8 +71,16 @@ def test_loop_trans_validate(monkeypatch):
...
+    # Pretend that the loop is of 'null' type
+    monkeypatch.setattr(loop, "_loop_type", "null")
+    with pytest.raises(TransformationError) as err:
+        trans.validate(loop)
+    assert ("Cannot apply a OMPParallelLoopTrans transformation to a "
+            "'null' loop" in str(err.value))
+    monkeypatch.undo()
+    # Break the contents of the loop
     loop.children = loop.pop_all_children()[0:1]
+    # Break the contents of the loop
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/loop_trans_test.py:83]: Removed

**Comment by sergisiso** [comment]
@arporter This is ready for next review, I added a section in the developers guide.

**Comment by sergisiso** [comment]
@arporter I pushed some commits to bring it up to date with master, but it is still Ready

**Comment by arporter** [comment]
Thanks for adding the new documentation - I think it's good. (I've taken the liberty of tweaking it slightly.) There were also 3 new warnings when building the reference guide but since one of those was my fault anyway I've also fixed those.
Apart from that, tests and examples are all OK with compilation.
All requested changes have been made.

**Review by arporter — CHANGES_REQUESTED** [review]
I was going to merge this time but then I did a 'git submodule update' and now some of the tests and examples fail when compilation is enabled, e.g.:

    gfortran  -I../../../external/dl_esm_inf/finite_difference/src -I../../../external/FortCL/src -c psy_gen.f90
    psy_gen.f90:156:28:

    156 |         cu_fld%device_ptr = create_rw_buffer(size_in_bytes)
        |                            1
    Error: Cannot convert INTEGER(8) to TYPE(c_ptr) at (1)

Do these work for you @sergisiso?

**Comment by sergisiso** [comment]
@arporter Which version of dl_esm_inf do you get after the module update? This issue seems caused by the latest updates I did to dl_esm_inf but this shouldn't be on PSyclone master or this PR (they come with #1175)

**Comment by sergisiso** [comment]
@arporter I think I inadvertently end up with the module update in this PR, I reverted it now. Thanks for checking this. You should be able to run the compile tests now I think.

**Comment by arporter** [comment]
Thanks for fixing Sergi. Tests, examples and notebooks are all OK for me now in Python 2 and Python 3. I'll force the merge to master since, as you pointed out, making 3.5 optional doesn't stop GHA from killing all the other tests :-(

**Comment by arporter** [comment]
And now py 3.5 magically works again! I've just checked and the version of `notebook` being installed in both cases is the same.

**Review by arporter — APPROVED** [review]

**Pull request merged**